pylace/lace/engine.py

```diff
@@ -39,66 +39,113 @@ def __init__(self, df: pl.DataFrame, linkage: "np.ndarray", figure=None):
-            The path to the source data file.
-        codebook: path-like, optional
-            Path to the codebook. If ``None`` (default), a codebook is inferred.
-        n_states: usize
-            The number of states (independent Markov chains). default is 16.
-        id_offset: int
+        dataframe: pd.DataFrame or pl.DataFrame
+            DataFrame with relevant data.
+        codebook_builder: CodebookBuilder or PathLike or str, optional
+            Codebook builder which can load codebook from file or generate one
+            from data. See ``CodebookBuilder``.
+        n_states: int, optional
+            The number of states (independent Markov chains).
+        id_offset: int, optional
             An offset for renaming states in the metadata. Used when training a
             single engine on multiple machines. If one wished to split an
             8-state ``Engine`` run on to two machine, one may run a 4-state
             ``Engine`` on the first machine, then a 4-state ``Engine`` on the
             second machine with ``id_offset=4``. The states within two metadata
             files may be merged by copying without name collisions.
-        rng_seed: int
-            Random number generator seed
-        source_type: str, optional
-            The type of the source file. If ``None`` (default) the type is
-            inferred from the file extension.
-        cat_cutoff: int, optional
-            The maximum integer value an all-integer column takes on at which
-            it is considered count type.
-        no_hypers: bool
-            If ``True``, hyper priors, and prior parameter inference will be
-            disabled
+        rng_seed: int, optional
+            Random number generator seed.
 
 
         Examples
         --------
-        Load an Engine from metadata
+        Create a new ``Engine`` from a DataFrame
 
-        >>> from lace import Engine  # doctest: +SKIP
-        >>> engine = Engine(metadata="metadata.lace")  # doctest: +SKIP
+        >>> from lace import Engine
+        >>> import polars as pl
+        >>> df = pl.DataFrame({
+        >>>    "list_a": [[1, 2, 3], [5, 5]],
+        >>>    "list_b": [[1, 2, 3, 8], [5, 1, 1]]
+        >>> })
+        >>> engine = Engine.from_dataframe(df)
+
+        Create a new ``Engine`` with specific codebook inference rules
+        >>> from lace import Engine, CodebookBuilder
+        >>> import polars as pl
+        >>> df = pl.DataFrame({
+        >>>    "list_a": [[1, 2, 3], [5, 5]],
+        >>>    "list_b": [[1, 2, 3, 8], [5, 1, 1]]
+        >>> })
+        >>> engine = Engine.from_dataframe(df, CodebookBuilder.infer(
+        >>>     cat_cutoff=2,
+        >>> ))
 
-        Create a new Engine with default codebook. The start state is drawn from
-        the probabilistic cross-categorization prior.
+        """
+        if isinstance(dataframe, pd.DataFrame):
+            dataframe = pl.DataFrame._from_pandas(dataframe)
```

Comment:
Does this function convert the pandas index into a column in the polars series or is it removed?

Reply:
Fixed.